File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        - add ( when openCount < n 
        - add ) when closedCount < openCount
        - valid if openCount == closedCount == n
        """

        # hold parentheses 
        stack = []
        
        # list of valid parentheses combinations 
        result = []

        # function passes in open n and closed n 
        def backtrack(openCount, closeCount):

            # base case 
            # if open n == close n == n
            if openCount == closeCount == n:
                # take all characters in a string 
                # take all characters in the stack and join them into an empty string and append to result list 
                result.append("".join(stack))

                # return base case 
                return

            # check if open count is less than end 
            # add just an open parenthesis to stack
            if openCount < n:
                stack.append("(")

                # recursively continue backtrack by incrementing open count by one and leave closed the same 
                backtrack(openCount + 1, closeCount)

                # pop the character just added to the stack 
                stack.pop()

            # make sure close count is less than the open count 
            if closeCount < openCount:
                # append the closing parenthesis to the stack
                stack.append(")")

                # call recursive backtrack and increment closing count by 1 and leave open the same 
                backtrack(openCount, closeCount + 1)

                # update stack by popping the character that was just added 
                stack.pop()

        # call backtrack function and start both open and closed as 0, 0
        backtrack(0, 0)

        # return results 
        return result

File: 0022-generate-parentheses/0022-generate-parentheses.java
import java.util.*;

class Solution {
    public List<String> generateParenthesis(int n) {

        // Stores all valid combinations
        List<String> answer = new ArrayList<>();

        // Temporary list used to build the current combination
        List<Character> result = new ArrayList<>();

        // Start backtracking with 0 open and 0 close parentheses
        backtrack(0, 0, n, result, answer);

        return answer;
    }

    private void backtrack(
        int openCount,
        int closeCount,
        int n,
        List<Character> result,
        List<String> answer
    ) {

        // If we have used n opening AND n closing parentheses,
        // we have created a complete valid combination.
        if (openCount == n && closeCount == n) {

            // Convert the characters into a String
            StringBuilder sb = new StringBuilder();

            for (char c : result) {
                sb.append(c);
            }

            answer.add(sb.toString());

            return;
        }

        // We can add an opening parenthesis as long as
        // we haven't used all n opening parentheses.
        if (openCount < n) {
            result.add('(');

            // Recursively continue building the combination
            backtrack(
                openCount + 1,
                closeCount,
                n,
                result,
                answer
            );

            // Remove the '(' so we can try another possibility.
            // This is the "backtrack" part.
            result.remove(result.size() - 1);
        }

        // We can only add a closing parenthesis if there
        // are currently more '(' than ')' .
        //
        // This prevents invalid combinations like:
        // ")("
        // "())("
        if (closeCount < openCount) {
            result.add(')');

            // Recursively continue building the combination
            backtrack(
                openCount,
                closeCount + 1,
                n,
                result,
                answer
            );

            // Remove the ')' to backtrack
            result.remove(result.size() - 1);
        }
    }
}